agent/grpc-external/services/resource/write.go

```diff
@@ -0,0 +1,144 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/lib/retry"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+func (s *Server) Write(ctx context.Context, req *pbresource.WriteRequest) (*pbresource.WriteResponse, error) {
+	if err := validateWriteRequiredFields(req); err != nil {
+		return nil, err
+	}
+
+	reg, err := s.resolveType(req.Resource.Id.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the user sent the correct type of data.
+	if !req.Resource.Data.MessageIs(reg.Proto) {
+		got := strings.TrimPrefix(req.Resource.Data.TypeUrl, "type.googleapis.com/")
+
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"resource.data is of wrong type (expected=%q, got=%q)",
+			reg.Proto.ProtoReflect().Descriptor().FullName(),
+			got,
+		)
+	}
+
+	var result *pbresource.Resource
+	err = s.retryCAS(ctx, req.Resource.Version, func() error {
+		input := clone(req.Resource)
+
+		// Read with EventualConsistency because we'll already automatically retry on
+		// CAS failure, so it's not worth the latency penalty of StrongConsistency.
+		existing, err := s.Backend.Read(ctx, storage.EventualConsistency, input.Id)
+		switch {
+		case err == nil:
+			if input.Id.Uid == "" {
+				input.Id.Uid = existing.Id.Uid
+			}
+
+			if input.Version == "" {
+				input.Version = existing.Version
+			} else if input.Version != existing.Version {
+				// Although the storage backend will check the version on-write, we check
+				// it here too to make sure we don't carry over statuses from the wrong
+				// version.
+				return storage.ErrCASFailure
+			}
+			// TODO: Carry over the statuses here.
+
+		case errors.Is(err, storage.ErrNotFound):
+			input.Id.Uid = ulid.Make().String()
+			// TODO: Prevent setting statuses in this endpoint.
+
```

Comment:
Maybe add a comment here that this is a "create" hence generating and assigning a Uid?

Reply:
Sure thing! 👍🏻